### ISSUE TITLE
False positive with type_repetition_in_bounds and generics

### ISSUE THREAD

**Issue by taiki-e** [rust-lang/rust-clippy#4323]
`type_repetition_in_bounds` seems missed generics in types.

```rust
#![allow(dead_code)]
#![warn(clippy::type_repetition_in_bounds)]

pub struct Foo<A>(A);

pub struct Bar<A, B> {
    a: Foo<A>,
    b: Foo<B>,
}

impl<A, B> Unpin for Bar<A, B>
where
    Foo<A>: Unpin,
    Foo<B>: Unpin, //~ WARN this type has already been used as a bound predicate
{
}
```

**Comment by flip1995** [comment]
cc @xd009642 Any idea why this happens? Seems like the hashing of the types is missing the generics.

**Comment by xd009642** [comment]
I'll have a look tonight but that seems the most likely scenario

**Comment by xd009642** [comment]
So I can't see any parameters in `Ty` or `TyKind` for generic types associated with the type. There's 1rustc::hir::WhereBoundPredicate::bounded_generic_params` which maybe should be hashed? Provided that it's just the generics used in each bounded type. I'll have a play and see what I can do

**Comment by flip1995** [comment]
> `rustc::hir::WhereBoundPredicate::bounded_generic_params` which maybe should be hashed

That sounds promising. Thanks for taking care of this!

**Comment by ThibsG** [comment]
I cannot reproduce this warning. Am I doing something wrong or the issue has gone?

**Comment by taiki-e** [comment]
I don't think this has been fixed: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7578f98b9cd871f36e8461ed44e744e7)

**Comment by ThibsG** [comment]
Indeed, I cannot trigger this on my local Clippy test, I will figure out why

**Comment by taiki-e** [comment]
> I cannot trigger this on my local Clippy test

I guess it's one of the following:
* `compiletest` may ignore the warning, so you might be able to reproduce this by using `#![deny(...)]` instead of `#![warn(...)]`.
* This has been fixed in the `master` but not yet released.

**Comment by ThibsG** [comment]
Yes I just fixed it, something went wrong with `deny/warn` and my eyes.
